addons/im_livechat/models/chatbot_script_step.py

```diff
@@ -0,0 +1,199 @@
+    _name = 'im_livechat.chatbot.script_step'
+    _description = 'Chatbot Script Step'
+    _order = 'sequence, id'
+
+    message = fields.Text(string='Message', required=True)
+    sequence = fields.Integer(string='Sequence')
+    chatbot_id = fields.Many2one(
+        'im_livechat.chatbot.script', string='Chatbot', required=True, ondelete='cascade')
+    step_type = fields.Selection([
+        ('text', 'Text'),
+        ('question_selection', 'Question'),
+        ('question_email', 'Email'),
+        ('question_phone', 'Phone'),
+        ('forward_operator', 'Forward to Operator'),
+    ], default='text', required=True)
+    # answers
+    answer_ids = fields.One2many(
+        'im_livechat.chatbot.script_question_answer', 'step_id',
+        copy=True, string='Answers')
+    triggering_answer_ids = fields.Many2many(
+        'im_livechat.chatbot.script_question_answer', 'chatbot_script_chatbot_script_question_answer_rel',
+        domain="[('step_id.sequence', '<', sequence)]",
+        copy=False,  # copied manually, see im_livechat.chatbot.script#copy
+        string='Only If', help='Show this step only if all of these answers have been selected.')
+    # forward-operator specifics
+    is_forward_operator_child = fields.Boolean(compute='_compute_is_forward_operator_child')
+
+    @api.onchange('sequence')
+    def onchange_sequence(self):
+        """ Unlink the triggering answers that comes from the following steps after a sequence change """
+        if self.triggering_answer_ids:
+            update_command = [Command.unlink(answer.id) for answer in self.triggering_answer_ids
+                              if answer.step_id.sequence >= self.sequence]
+            if update_command:
+                self.write({'triggering_answer_ids': update_command})
+
+    def _get_parent(self):
+        """ Returns the first preceding step that matches either the triggering answers
+         or the possible answers the user can select """
+        self.ensure_one()
+
+        if not self.chatbot_id.ids:
+            return False
+
+        if not self.triggering_answer_ids:
+            return False
+        else:
+            domain = [
+                ('chatbot_id', 'in', self.chatbot_id.ids),  # Used ids instead of id to avoid testing on NewId
+                ('step_type', 'in', ['forward_operator', 'question_selection']),
+                ('sequence', '<', self.sequence),
+            ]
+            answer_domain = []
+            for answer in self.triggering_answer_ids:
+                answer_domain = expression.AND([answer_domain, [('triggering_answer_ids', 'in', answer.ids)]])
+            domain = expression.AND([domain, expression.OR([
+                answer_domain, [('answer_ids', 'in', self.triggering_answer_ids.ids)]])
+            ])
+        return self.search(domain, order='sequence desc', limit=1)
+
+    @api.depends('sequence', 'triggering_answer_ids')
+    def _compute_is_forward_operator_child(self):
```

Comment:
If I understand well, purpose is to know that you are the child of a "forward to operator" step, in the same script. Current implementation seems sub-optimized (doing searches in loop). As scripts should not be that much complex, isn't it easier to simply check all previous steps of the current script and build the tree manually ? That way we avoid search in a while loop, better do it in python if we can as search is more costly than a single browse and read.

Reply:
Better with a single browse prefetching data indeed :)